File: routeMap/app/src/main/java/ru/android/zheka/gmapexample1/TimerService.java
package ru.android.zheka.gmapexample1;

import java.util.Iterator;
import java.util.concurrent.ConcurrentLinkedDeque;

import ru.android.zheka.db.Config;
import ru.android.zheka.db.DbFunctions;
import ru.zheka.android.timer.Recievable;
import android.app.IntentService;
import android.content.Intent;

public class TimerService extends IntentService{
	public TimerService() {
		super("TimerService");
		System.out.println("TimerService() is called");
		if (mListners==null) mListners = new ConcurrentLinkedDeque<Recievable>();
	}/*
	static TimerService instance = null;
	static TimerService getInstance(){
		if (instance==null) instance = new TimerService();
		return instance;
	}*/

	static boolean interrupted = true;
	static final String BROADCAST_ACTION = "ru.android.zheka.gmapexample1.Timer";
	static ConcurrentLinkedDeque<Recievable> mListners = new ConcurrentLinkedDeque<Recievable>();
	
	@Override
	protected void onHandleIntent(Intent intent) {
		interrupted = false;
		Config config = (Config) DbFunctions.getModelByName(DbFunctions.DEFAULT_CONFIG_NAME
				, Config.class);
		Integer ms=0;
		try{ms = new Integer(config.tenMSTime);		
		}catch(NumberFormatException e){
			e.printStackTrace();
			return;
		}
		intent.setAction(TimerService.BROADCAST_ACTION);
		long start = System.nanoTime();
		while(ms!=0 
				&& !interrupted){
			try {
				Thread.currentThread().join(ms);
			} catch (InterruptedException e) {
				interrupted = true;
			}
			long time = (System.nanoTime()-start)/1000000;
			System.out.println("current time(ms): "+time);
			sendBroadcast(intent);
			config = (Config) DbFunctions.getModelByName(DbFunctions.DEFAULT_CONFIG_NAME
					, Config.class);
			try{ ms = new Integer(config.tenMSTime);		
			}catch(NumberFormatException e){
				e.printStackTrace();
				return;
			}
		}
		interrupted = true;// finish see doc 
	}
	public boolean add(Recievable listner){
		return mListners.add(listner);
	}
	public boolean remove(Recievable listner){
		return mListners.remove(listner);
	}
	public void sendBroadcast(Intent intent){
		System.out.println("hello from sendBroadcast, mListners is "+mListners);
		for (Iterator iterator = mListners.iterator(); iterator.hasNext();) {
			Recievable l = (Recievable) iterator.next();
			System.out.println("recivabale:"+l);
			l.onReceive(getApplicationContext(), intent);			
		}
	}
}

File: routeMap/app/src/main/java/ru/android/zheka/gmapexample1/TimerService.kt
package ru.android.zheka.gmapexample1

import android.app.IntentService
import android.content.Intent
import android.os.Build
import androidx.annotation.RequiresApi
import ru.android.zheka.db.Config
import ru.android.zheka.db.DbFunctions
import ru.zheka.android.timer.Recievable
import java.util.concurrent.ConcurrentLinkedDeque

@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
class TimerService : IntentService("TimerService") {
    override fun onHandleIntent(intent: Intent) {
        interrupted = false
        var config = DbFunctions.getModelByName(DbFunctions.DEFAULT_CONFIG_NAME
                , Config::class.java) as Config
        var ms:Int
        try {
            ms = config.tenMSTime?.toInt()?:0
        } catch (e: NumberFormatException) {
            e.printStackTrace()
            return
        }
        intent.action = BROADCAST_ACTION
        val start = System.nanoTime()
        while (ms != 0
                && !interrupted) {
            try {
                Thread.currentThread().join(ms.toLong())
            } catch (e: InterruptedException) {
                interrupted = true
            }
            val time = (System.nanoTime() - start) / 1000000
            println("current time(ms): $time")
            sendBroadcast(intent)
            config = DbFunctions.getModelByName(DbFunctions.DEFAULT_CONFIG_NAME
                    , Config::class.java) as Config
            try {
                ms = config.tenMSTime?.toInt()?:0
            } catch (e: NumberFormatException) {
                e.printStackTrace()
                return
            }
        }
        interrupted = true // finish see doc
    }

    fun add(listner: Recievable): Boolean {
        return mListners!!.add(listner)
    }

    fun remove(listner: Recievable): Boolean {
        return mListners!!.remove(listner)
    }

    override fun sendBroadcast(intent: Intent) {
        println("hello from sendBroadcast, mListners is $mListners")
        val iterator: Iterator<*> = mListners!!.iterator()
        while (iterator.hasNext()) {
            val l = iterator.next() as Recievable
            println("recivabale:$l")
            l.onReceive(applicationContext, intent)
        }
    }

    companion object {
        var interrupted = true
        const val BROADCAST_ACTION = "ru.android.zheka.gmapexample1.Timer"
        var mListners: ConcurrentLinkedDeque<Recievable>? = ConcurrentLinkedDeque()
    }

    init {
        println("TimerService() is called")
        if (mListners == null) mListners = ConcurrentLinkedDeque()
    } /*
	static TimerService instance = null;
	static TimerService getInstance(){
		if (instance==null) instance = new TimerService();
		return instance;
	}*/
}